processor/src/operations/decorators/mod.rs

```diff
@@ -7,9 +11,79 @@ impl Process {
     // DEBUGGING
     // --------------------------------------------------------------------------------------------
 
-    /// TODO: implement
-    pub fn op_debug(&self) -> Result<(), ExecutionError> {
-        unimplemented!();
+    /// Prints out debugging information based on options passed.
+    pub fn op_debug(&mut self, options: DebugOptions) -> Result<(), ExecutionError> {
+        match options {
+            DebugOptions::All => {
+                self.print_stack(None);
+                self.print_mem(None, None);
+            }
+            DebugOptions::Stack(n) => self.print_stack(n),
+            DebugOptions::Mem(n, m) => self.print_mem(n, m),
+            DebugOptions::Local(_) => self.print_local(),
+        }
+
+        Ok(())
+    }
+
+    // DEBUG HELPER
+    // ---------------------------------------------------------------------------------
+
+    /// Prints stack information for debugging.
+    /// If n is passed, this prints the top n items in the stack.
+    fn print_stack(&self, n: Option<usize>) {
+        let states = self.stack.get_values(n);
+        for (i, value) in states.iter().enumerate() {
+            info!("stack [ index: {:?}, value: {:?} ]", i, value.as_int());
+        }
+    }
+
+    /// Create the display string for printing a word.
+    /// Prints <empty> if no value exists.
+    fn fmt_word(word: Option<Word>) -> String {
+        match word {
+            Some(v) => format!(
+                "[{}]",
+                v.iter()
+                    .map(|v| v.as_int().to_string())
+                    .collect::<Vec<String>>()
+                    .join(", ")
+            ),
+            None => String::from("<empty>"),
+        }
+    }
+
+    /// Print the local variable that fmp pointer is referring to in memory.
+    fn print_local(&self) {
+        let local = self.memory.get_value(self.system.fmp().as_int());
+        info!("local [ {} ]", Process::fmt_word(local));
+    }
+
+    /// Print memory with optional starting and ending addresses.
+    fn print_mem(&self, n: Option<u64>, m: Option<u64>) {
+        // Convert vec of words into options
+        let convert_vec = |v: Vec<(u64, Word)>| {
+            v.into_iter()
+                .map(|(k, v)| (k, Some(v)))
+                .collect::<Vec<(u64, Option<Word>)>>()
+        };
+        let values: Vec<(u64, Option<Word>)> = match (n, m) {
+            (Some(n), None) => {
+                let value = self.memory.get_value(n);
+                vec![(n, value)]
+            }
+            (Some(n), Some(m)) => convert_vec(self.memory.get_values(RangeInclusive::new(n, m))),
+            (None, None) => convert_vec(self.memory.get_all_values()),
+            _ => Vec::new(),
+        };
+
+        for (address, value) in values {
+            info!(
+                "mem [ address: {:016x}, value: {} ]",
+                address,
+                Process::fmt_word(value)
```

Comment:
We can probably don't need to have labels for each value. Maybe something like this could work:

```
0x0000000000000001: [1, 2, 3, 4]
0x0000000000000002: [5, 6, 7, 8]
```